src/components/Pagination/index.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { FC } from "react"
+
+interface PaginationProps {
+  page?: number | null
+  perPage?: number | null
+  totalItems: number
+  setPagination: (pagination: { page?: number; perPage?: number }) => void
+  onPageChange: (newPage: number) => void
+  maxVisiblePages?: number
+}
+
+const Pagination: FC<PaginationProps> = ({
+  page,
+  perPage,
+  totalItems,
+  setPagination,
+  onPageChange,
+  maxVisiblePages = 4,
+}) => {
+  if (!page || !perPage) return null
+
+  const totalPages = Math.ceil(totalItems / perPage)
+
+  const handlePageChange = (newPage: number) => {
+    setPagination({ page: newPage })
+    onPageChange(newPage)
+  }
+
+  const renderPageNumbers = () => {
+    const pageNumbers = []
+    const halfMaxVisiblePages = Math.floor(maxVisiblePages / 2)
+    let startPage = Math.max(page - halfMaxVisiblePages, 1)
+    let endPage = Math.min(startPage + maxVisiblePages - 1, totalPages)
```

Comment:
Looks to me `endPage` can be a const.
With a little refactoring you can do check from line 36 for start page right away and make it into a const as well.
I would recommend a very simple utility function we use in all of our projects. We call it `compute`:
```ts
const compute<T>(callback: () => T) {
    return callback()
}
```
It works similar to `useMemo` but has no dependency array so it's always called.
We use it to eliminate `let` variables since they make the code harder to read by changing their value based on conditions which breaks the flow of functions (in my opinion ofc).

it would work something like this:
```ts
const endPage = Math.min(startPage + maxVisiblePages - 1, totalPages)
const startPage = compute<number>(() => {
    const halfPoint = Math.max(page - halfMaxVisiblePages, 1)
    const lastPage = maxVisiblePages - 1
    return endPage - halfPoint < lastPage ? Math.max(endPage - maxVisiblePages + 1, 1) : halfPoint
})

// OR without compute but still no let
const endPage = Math.min(startPage + maxVisiblePages - 1, totalPages)
const startPage = endPage - startPage < maxVisiblePages - 1
    ? Math.max(endPage - maxVisiblePages + 1, 1)
    : Math.max(page - halfMaxVisiblePages, 1)
```

If you are not bothered by `let` leave it as-is. I would recommend to take an opportunity to test multiple approaches to the same problem. It will help learn more patterns you can use in the future.